File: 程序员面试金典第6版/面试题 16.06. Smallest Difference LCCI.cpp
class Solution {
public:
    int smallestDifference(vector<int>& a, vector<int>& b) {
        sort(a.begin(), a.end());
        sort(b.begin(), b.end());
        int i=0, j=0;
        long res = INT_MAX;
        while(i < a.size() && j < b.size()) {
            if(a[i] < b[j]) {
                res = min(res, ((long) b[j] - (long) a[i])); // 括号不能省
                i++;
            } else {
                res = min(res, ((long) a[i] - (long) b[j]));
                j++;
            }
        }
        return res;
    }
};

File: 程序员面试金典第6版/面试题 16.06. Smallest Difference LCCI.java
// 暴力法 需要处理溢出 而且超时
// class Solution {
//     public int smallestDifference(int[] a, int[] b) {
//         long res = Long.MAX_VALUE;
//         for(int i=0; i<a.length; i++) {
//             for(int j=0; j<b.length; j++) {
//                 long diff = b[j]-a[i];
//                 if(diff < 0) diff = -diff;
//                 res = Math.min(res, diff);
//             }
//         }
//         return (int) res;
//     }
// }

// two pointers

class Solution {
    public int smallestDifference(int[] a, int[] b) {
        Arrays.sort(a);
        Arrays.sort(b);
        int i=0, j=0;
        long res = Integer.MAX_VALUE;
        while(i < a.length && j < b.length) {
            if(a[i] < b[j]) {
                res = Math.min(res, (long) b[j] - (long) a[i]);
                i++;
            } else {
                res = Math.min(res, (long) a[i] - (long) b[j]);
                j++;
            }
        }
        return (int) res;
    }
}